lib/spack/spack/cmd/diff.py

```diff
@@ -0,0 +1,169 @@
+    into a json object for the user to save. We return an object that shows
+    the differences, intersection, and names for a pair of specs a and b.
+
+    Arguments:
+        a (spec): the first spec to compare
+        b (spec): the second spec to compare
+        a_name (str): the name of spec a
+        b_name (str): the name of spec b
+        to_string (bool): return an object that can be json dumped
+    """
+    from spack.solver.asp import SpackSolverSetup
+
+    # Prepare a solver setup to parse differences
+    setup = SpackSolverSetup()
+
+    a_facts = set(to_tuple(t) for t in setup.spec_clauses(a))
+    b_facts = set(to_tuple(t) for t in setup.spec_clauses(b))
+
+    # We want to present them to the user as simple key: values
+    intersect = list(a_facts.intersection(b_facts))
+    spec1_not_spec2 = list(a_facts.difference(b_facts))
+    spec2_not_spec1 = list(b_facts.difference(a_facts))
+
+    # We want to show what is the same, and then difference for each
+    return {
+        "intersect": flatten(intersect) if to_string else intersect,
+        "a_not_b": flatten(spec1_not_spec2) if to_string else spec1_not_spec2,
+        "b_not_a": flatten(spec2_not_spec1) if to_string else spec2_not_spec1,
+        "a_name": a.name_version_build_hash,
+        "b_name": b.name_version_build_hash,
+    }
+
+
+def to_tuple(asp_function):
+    """Prepare tuples of objects. If we need to save to json, convert to strings
+    See https://gist.github.com/tgamblin/83eba3c6d27f90d9fa3afebfc049ceaf
+    """
+    args = []
+    for arg in asp_function.args:
+        if isinstance(arg, str):
+            args.append(arg)
+            continue
+        args.append("%s(%s)" % (type(arg).__name__, str(arg)))
+    return tuple([asp_function.name] + args)
+
+
+def flatten(tuple_list):
+    """Given a list of tuples, convert into a list of key: value tuples (so
+    we are squashing whatever is after the first index into one string for
+    easier parsing in the interface
+    """
+    updated = []
+    for item in tuple_list:
+        updated.append([item[0], " ".join(item[1:])])
+    return updated
+
+
+def print_difference(diffset, diff_type="all"):
+    """Given a diffset and a user preference (e.g., print versions or print all)
+    print a tabular, organized version of the diffset
+    """
+    # Cut out early if we don't have any differences!
+    if not diffset:
+        print("No differences\n")
+        return
+
+    # Sort by name so they are grouped together
+    sorted_diffset = sorted(diffset, key=operator.itemgetter(0))
+
+    # Keep rows to print
+    rows = []
+
+    # Always print a new category
+    category = None
+    for entry in sorted_diffset:
+        if diff_type == "all" or diff_type == entry[0]:
+
+            # If we have a new category, create a new section
+            if category != entry[0]:
+                category = entry[0]
+                rows.append(bold(category.upper()))
+            rows.append(entry[1])
```

Comment:
Can you add some indentation here so the different values are offset from the bolded headings?  It'll make it a bit more readable, especially if there are a lot of differences.

```suggestion
            rows.append('  {0}'.format(entry[1]))
```